node/src/components/storage/disjoint_sequences.rs

```diff
@@ -0,0 +1,317 @@
+        // The index of a `Sequence` to be removed due to the insertion of `value` causing two
+        // consecutive sequences to become contiguous.
+        let mut maybe_removal_index = None;
+        while let Some((index, sequence)) = iter_mut.next() {
+            match sequence.try_insert(value) {
+                InsertOutcome::ExtendedHigh | InsertOutcome::AlreadyInSequence => {
+                    // We should exit the loop, and we don't need to add a new sequence; we only
+                    // need to check for merges of sequences when we get `ExtendedLow` since we're
+                    // iterating the sequences from high to low.
+                    maybe_insertion_index = None;
+                    break;
+                }
+                InsertOutcome::TooHigh => {
+                    // We should exit the loop and we need to add a new sequence at this index.
+                    maybe_insertion_index = Some(index);
+                    break;
+                }
+                InsertOutcome::TooLow => {
+                    // We need to add a new sequence immediately after this one if this is the last
+                    // sequence.  Continue iterating in case this is not the last sequence.
+                    maybe_insertion_index = Some(index + 1);
+                }
+                InsertOutcome::ExtendedLow => {
+                    // We should exit the loop, and we don't need to add a new sequence.
+                    maybe_insertion_index = None;
+                    // If the next sequence is now contiguous with this one, update this one's low
+                    // value and set the next sequence to be removed.
+                    if let Some((next_index, next_sequence)) = iter_mut.peek() {
+                        if next_sequence.high + 1 == sequence.low {
+                            sequence.low = next_sequence.low;
+                            maybe_removal_index = Some(*next_index);
+                        }
+                    }
+                    break;
+                }
+            };
+        }
+
+        if let Some(index_to_insert) = maybe_insertion_index {
+            let _ = self.sequences.insert(index_to_insert, Sequence::new(value));
+        }
+
+        if let Some(index_to_remove) = maybe_removal_index {
+            let _ = self.sequences.remove(index_to_remove);
+        }
+
+        trace!(%self, "current state of disjoint sequences");
+    }
+
+    /// Inserts multiple values produced by the given interator.
+    pub(super) fn extend<T>(&mut self, iter: T)
+    where
+        T: IntoIterator<Item = u64>,
+    {
+        iter.into_iter().for_each(|height| self.insert(height))
+    }
+
+    /// Returns the `low` value from the highest sequence, or `u64::MAX` if there are no sequences.
+    pub(super) fn highest_sequence_low_value(&self) -> u64 {
+        self.sequences
+            .first()
+            .map(|sequence| sequence.low)
+            .unwrap_or(u64::MAX)
+    }
```

Comment:
Should this maybe return an `Option<u64>`? I feel like by substituting the lack of sequences with `u64::MAX`, we're setting up a trap for ourselves where we'll have to remember to check if the value is reasonable at every call site - and if we return an `Option`, the compiler will watch us.

Reply:
Well, we can maybe compromise here.  The only production callsite is the handling of `GetLowestContiguousBlockHeight` (which will have been made indirectly from a JSON-RPC request), and given that returning an error here would be a fatal error for the node, I think we'd want to return _some_ value.

I'll add an error log here instead since that should draw attention to there being a problem if it ever prints.

Reply:
Fair point - I'll do that.